### PR TITLE
feat(RDS): rds instance support restore

### LINKED ISSUE #2124: [RDS] Support use backup file to create new RDS instacnce.
### Proposal

Hope  support use backup file to create new RDS instacnce

https://support.huaweicloud.com/api-rds/rds_09_0008.html#section4

### PR DESCRIPTION
**Pull request by houpeng80**
**What this PR does / why we need it**:
  rds instance support restore
**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:
  
**Release note**:

```release-note
  rds instance support restore
```

## PR Checklist

* [ ] Tests added/passed.
* [ ] Documentation updated.
* [ ] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST='./huaweicloud/services/acceptance/rds/' TESTARGS='-run TestAccRdsInstance_restore_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/rds/ -v -run TestAccRdsInstance_restore_ -timeout 360m -parallel 4
=== RUN   TestAccRdsInstance_restore_mysql
=== PAUSE TestAccRdsInstance_restore_mysql
=== RUN   TestAccRdsInstance_restore_sqlserver
=== PAUSE TestAccRdsInstance_restore_sqlserver
=== RUN   TestAccRdsInstance_restore_pg
=== PAUSE TestAccRdsInstance_restore_pg
=== CONT  TestAccRdsInstance_restore_mysql
=== CONT  TestAccRdsInstance_restore_pg
=== CONT  TestAccRdsInstance_restore_sqlserver
--- PASS: TestAccRdsInstance_restore_pg (2097.87s)
--- PASS: TestAccRdsInstance_restore_mysql (2522.02s)
--- PASS: TestAccRdsInstance_restore_sqlserver (2690.34s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/rds       5264.663s
```

### CODE REVIEW

File: docs/resources/rds_instance.md [hunk 1]
```diff
@@ -188,6 +188,9 @@ The following arguments are supported:
 
 * `volume` - (Required, List) Specifies the volume information. Structure is documented below.
 
+* `restore` - (Optional, List, ForceNew) Specifies the restoration information. It only supported restore to postpaid
+  instance. Structure is documented below.
```
Comment (Jason-Zhang9309) [inline docs/resources/rds_instance.md]: missing `Changing this ...`

File: docs/resources/rds_instance.md [hunk 2]
```diff
@@ -281,6 +284,15 @@ The `volume` block supports:
...
 
+The `restore` block supports:
+
+* `instance_id` - (Required, String, ForceNew) Specifies the DB instance ID.
+
+* `backup_id` - (Required, String, ForceNew) Specifies the ID of the backup used to restore data.
+
+* `database_name` - (Optional, Map, ForceNew) Specifies the database to be restored. This parameter applies only to
+  Microsoft SQL Server databases.
+
```
Comment (Jason-Zhang9309) [inline docs/resources/rds_instance.md]: put this at another place, this looks like `disk_encryption_id` is  in the restore block

File: docs/resources/rds_instance.md [hunk 3]
```diff
@@ -281,6 +284,15 @@ The `volume` block supports:
   Changing this parameter will create a new resource. For details about volume types, see
   [DB Instance Storage Types](https://support.huaweicloud.com/intl/en-us/productdesc-rds/rds_01_0020.html).
 
+The `restore` block supports:
+
+* `instance_id` - (Required, String, ForceNew) Specifies the DB instance ID.
```
Comment (Jason-Zhang9309) [inline docs/resources/rds_instance.md]: missing Changing this ...

File: huaweicloud/services/rds/resource_huaweicloud_rds_instance.go [hunk 4]
```diff
@@ -959,6 +991,22 @@ func buildRdsInstanceUnchangeableParam(d *schema.ResourceData) *instances.Unchan
...
+		restorePoint = new(instances.RestorePoint)
+		restorePoint.InstanceId = restorePointRaw[0].(map[string]interface{})["instance_id"].(string)
+		restorePoint.BackupId = restorePointRaw[0].(map[string]interface{})["backup_id"].(string)
+		databaseName := restorePointRaw[0].(map[string]interface{})["database_name"].(map[string]interface{})
+		restorePoint.DatabaseName = utils.ExpandToStringMap(databaseName)
+
+		restorePoint.Type = "backup"
+	}
+	return restorePoint
+}
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/rds/resource_huaweicloud_rds_instance.go]:
```suggestion
func buildRdsInstanceRestorePoint(d *schema.ResourceData) *instances.RestorePoint {
	var restorePoint instances.RestorePoint
	restorePointRaw := d.Get("restore").([]interface{})

	if len(restorePointRaw) == 1 {
                v := restorePointRaw[0].(map[string]interface{})
		restorePoint.InstanceId = v["instance_id"].(string)
		restorePoint.BackupId = v["backup_id"].(string)
		databaseName := v["database_name"].(map[string]interface{})
		restorePoint.DatabaseName = utils.ExpandToStringMap(databaseName)

		restorePoint.Type = "backup"
	}
	return &restorePoint
}
```

**Comment by houpeng80** [comment]
make testacc TEST='./huaweicloud/services/acceptance/rds/' TESTARGS='-run TestAccRdsInstance_restore_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/rds/ -v -run TestAccRdsInstance_restore_ -timeout 360m -parallel 4
=== RUN   TestAccRdsInstance_restore_mysql
=== PAUSE TestAccRdsInstance_restore_mysql
=== RUN   TestAccRdsInstance_restore_sqlserver
=== PAUSE TestAccRdsInstance_restore_sqlserver
=== RUN   TestAccRdsInstance_restore_pg
=== PAUSE TestAccRdsInstance_restore_pg
=== CONT  TestAccRdsInstance_restore_mysql
=== CONT  TestAccRdsInstance_restore_sqlserver
=== CONT  TestAccRdsInstance_restore_pg
--- PASS: TestAccRdsInstance_restore_pg (2147.12s)
--- PASS: TestAccRdsInstance_restore_mysql (2464.57s)
--- PASS: TestAccRdsInstance_restore_sqlserver (2749.31s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/rds       2749.366s

File: docs/resources/rds_instance.md [hunk 5]
```diff
@@ -293,6 +293,20 @@ The `volume` block supports:
   + **15**
   + **20**
 
+  Changing this parameter will create a new resource. For details about volume types, see
+  [DB Instance Storage Types](https://support.huaweicloud.com/intl/en-us/productdesc-rds/rds_01_0020.html).
```
Comment (ShiChangkuo) [inline docs/resources/rds_instance.md]: please move this line after the description of `type`

File: docs/resources/rds_instance.md [hunk 6]
```diff
@@ -293,6 +293,20 @@ The `volume` block supports:
...
+
+The `restore` block supports:
+
+* `instance_id` - (Required, String, ForceNew) Specifies the DB instance ID. Changing this parameter will create
```
Comment (ShiChangkuo) [inline docs/resources/rds_instance.md]:
```suggestion
* `instance_id` - (Required, String, ForceNew) Specifies the source DB instance ID. Changing this parameter will create
```

File: huaweicloud/services/rds/resource_huaweicloud_rds_instance.go [hunk 7]
```diff
@@ -363,10 +390,15 @@ func resourceRdsInstanceCreate(ctx context.Context, d *schema.ResourceData, meta
 		Volume:              buildRdsInstanceVolume(d),
 		Ha:                  buildRdsInstanceHaReplicationMode(d),
 		UnchangeableParam:   buildRdsInstanceUnchangeableParam(d),
+		RestorePoint:        buildRdsInstanceRestorePoint(d),
 	}
 
 	// PrePaid
 	if d.Get("charging_mode") == "prePaid" {
+		if createOpts.RestorePoint != nil {
```
Comment (ShiChangkuo) [inline huaweicloud/services/rds/resource_huaweicloud_rds_instance.go]: how about adding `ConflictsWith: []string{"period"}` in the schema definition of **restore**?

Comment (houpeng80) [reply huaweicloud/services/rds/resource_huaweicloud_rds_instance.go]: done

File: docs/resources/rds_instance.md (hunk 6 above)
Comment (houpeng80) [reply docs/resources/rds_instance.md]: done

File: docs/resources/rds_instance.md (hunk 5 above)
Comment (houpeng80) [reply docs/resources/rds_instance.md]: done

File: docs/resources/rds_instance.md (hunk 2 above)
Comment (houpeng80) [reply docs/resources/rds_instance.md]: done

File: huaweicloud/services/rds/resource_huaweicloud_rds_instance.go (hunk 4 above)
Comment (houpeng80) [reply huaweicloud/services/rds/resource_huaweicloud_rds_instance.go]: fixed

**Pull request merged**